File: common/common-cpp/src/camera.cpp
#include <gpusims/camera.hpp>

#include <algorithm>
#include <cmath>

#include <glm/gtc/matrix_transform.hpp>
#include <imgui.h>
#include <nlohmann/json.hpp>

namespace gpusims {

namespace {
constexpr float kPitchLimit = 89.0f;  // degrees from horizontal
}

Camera::Camera() = default;

void Camera::setMode(Mode m) {
    mode_ = m;
}

void Camera::setOrientation(float yaw_deg, float pitch_deg) {
    yaw_deg_   = yaw_deg;
    pitch_deg_ = std::clamp(pitch_deg, -kPitchLimit, kPitchLimit);
}

void Camera::resetArcball() {
    orbit_yaw_   = 0.0f;
    orbit_pitch_ = 20.0f;
}

glm::vec3 Camera::forward() const {
    if (mode_ == Mode::FreeFly) {
        const float yaw   = glm::radians(yaw_deg_);
        const float pitch = glm::radians(pitch_deg_);
        return glm::normalize(glm::vec3(
            std::cos(pitch) * std::cos(yaw),
            std::sin(pitch),
            std::cos(pitch) * std::sin(yaw)));
    }
    return glm::normalize(target_ - position_);
}

glm::vec3 Camera::right() const {
    return glm::normalize(glm::cross(forward(), glm::vec3(0.0f, 1.0f, 0.0f)));
}

glm::vec3 Camera::up() const {
    return glm::normalize(glm::cross(right(), forward()));
}

void Camera::update(float dt, const CameraInputState& input) {
    switch (mode_) {
    case Mode::FreeFly: {
        if (input.mouse_right) {
            yaw_deg_   += input.mouse_dx * look_speed_;
            pitch_deg_ -= input.mouse_dy * look_speed_;
            pitch_deg_  = std::clamp(pitch_deg_, -kPitchLimit, kPitchLimit);
        }
        const float speed = move_speed_ * (input.shift_held ? boost_mul_ : 1.0f);
        glm::vec3 vel(0.0f);
        const glm::vec3 fwd = forward();
        const glm::vec3 rt  = right();
        if (input.key_w) vel += fwd;
        if (input.key_s) vel -= fwd;
        if (input.key_d) vel += rt;
        if (input.key_a) vel -= rt;
        if (input.key_e) vel += glm::vec3(0.0f, 1.0f, 0.0f);
        if (input.key_q) vel -= glm::vec3(0.0f, 1.0f, 0.0f);
        if (glm::length(vel) > 0.0f) {
            position_ += glm::normalize(vel) * speed * dt;
        }
        break;
    }
    case Mode::Arcball: {
        if (input.mouse_left) {
            orbit_yaw_   += input.mouse_dx * look_speed_;
            orbit_pitch_ -= input.mouse_dy * look_speed_;
            orbit_pitch_  = std::clamp(orbit_pitch_, -kPitchLimit, kPitchLimit);
        }
        if (input.scroll_dy != 0.0f) {
            orbit_distance_ *= std::pow(1.1f, -input.scroll_dy);
            orbit_distance_  = std::max(orbit_distance_, 0.001f);
        }
        const float yaw   = glm::radians(orbit_yaw_);
        const float pitch = glm::radians(orbit_pitch_);
        position_ = target_ + orbit_distance_ * glm::vec3(
            std::cos(pitch) * std::cos(yaw),
            std::sin(pitch),
            std::cos(pitch) * std::sin(yaw));
        break;
    }
    case Mode::Orbit: {
        orbit_yaw_ += orbit_speed_ * dt;
        if (orbit_yaw_ > 360.0f) orbit_yaw_ -= 360.0f;
        const float yaw   = glm::radians(orbit_yaw_);
        const float pitch = glm::radians(orbit_pitch_);
        position_ = target_ + orbit_distance_ * glm::vec3(
            std::cos(pitch) * std::cos(yaw),
            std::sin(pitch),
            std::cos(pitch) * std::sin(yaw));
        break;
    }
    }
}

glm::mat4 Camera::view() const {
    if (mode_ == Mode::FreeFly) {
        return glm::lookAt(position_, position_ + forward(), glm::vec3(0.0f, 1.0f, 0.0f));
    }
    return glm::lookAt(position_, target_, glm::vec3(0.0f, 1.0f, 0.0f));
}

glm::mat4 Camera::projection() const {
    // Vulkan clip-space convention: y is flipped relative to OpenGL.
    glm::mat4 p = glm::perspective(glm::radians(fov_deg_), aspect_, near_, far_);
    p[1][1] *= -1.0f;
    return p;
}

void Camera::drawImGui(const char* label) {
    ImGui::SetNextWindowPos(ImVec2(10.0f, 10.0f),    ImGuiCond_FirstUseEver);
    ImGui::SetNextWindowSize(ImVec2(320.0f, 420.0f), ImGuiCond_FirstUseEver);
    if (!ImGui::Begin(label)) {
        ImGui::End();
        return;
    }
    static const char* const mode_names[] = {"Free-fly", "Arcball", "Orbit"};
    int mode_int = static_cast<int>(mode_);
    if (ImGui::Combo("Mode", &mode_int, mode_names, IM_ARRAYSIZE(mode_names))) {
        setMode(static_cast<Mode>(mode_int));
    }
    ImGui::SeparatorText("Lens");
    ImGui::SliderFloat("FOV (deg)", &fov_deg_, 10.0f, 120.0f);
    ImGui::DragFloat("Near", &near_, 0.001f, 0.0001f, far_ - 0.001f);
    ImGui::DragFloat("Far",  &far_,  0.1f,   near_ + 0.001f, 100000.0f);
    ImGui::SeparatorText("Position");
    ImGui::DragFloat3("Position", &position_.x, 0.05f);
    if (mode_ == Mode::FreeFly) {
        ImGui::SeparatorText("Free-fly");
        ImGui::SliderFloat("Yaw (deg)",   &yaw_deg_,   -180.0f, 180.0f);
        ImGui::SliderFloat("Pitch (deg)", &pitch_deg_, -kPitchLimit, kPitchLimit);
        ImGui::DragFloat("Move speed", &move_speed_, 0.05f, 0.0f, 1000.0f);
        ImGui::DragFloat("Look speed", &look_speed_, 0.005f, 0.0f, 5.0f);
        ImGui::DragFloat("Boost mul",  &boost_mul_,  0.1f,  1.0f, 100.0f);
    } else {
        ImGui::SeparatorText("Arcball / Orbit");
        ImGui::DragFloat3("Target", &target_.x, 0.05f);
        ImGui::DragFloat("Distance", &orbit_distance_, 0.01f, 0.001f, 10000.0f);
        ImGui::SliderFloat("Yaw (deg)",   &orbit_yaw_,   -180.0f, 180.0f);
        ImGui::SliderFloat("Pitch (deg)", &orbit_pitch_, -kPitchLimit, kPitchLimit);
        if (mode_ == Mode::Orbit) {
            ImGui::DragFloat("Orbit speed (deg/s)", &orbit_speed_, 0.5f, -360.0f, 360.0f);
        }
    }
    ImGui::End();
}

void Camera::toJson(nlohmann::json& j) const {
    j = nlohmann::json{
        {"mode",        static_cast<int>(mode_)},
        {"position",    {position_.x, position_.y, position_.z}},
        {"yaw_deg",     yaw_deg_},
        {"pitch_deg",   pitch_deg_},
        {"target",      {target_.x, target_.y, target_.z}},
        {"orbit", {
            {"distance", orbit_distance_},
            {"yaw",      orbit_yaw_},
            {"pitch",    orbit_pitch_},
            {"speed",    orbit_speed_},
        }},
        {"lens", {
            {"fov_deg", fov_deg_},
            {"aspect",  aspect_},
            {"near",    near_},
            {"far",     far_},
        }},
        {"freefly", {
            {"move_speed", move_speed_},
            {"look_speed", look_speed_},
            {"boost_mul",  boost_mul_},
        }},
    };
}

void Camera::fromJson(const nlohmann::json& j) {
    mode_       = static_cast<Mode>(j.value("mode", static_cast<int>(Mode::FreeFly)));
    auto pos    = j.value("position", std::vector<float>{0.0f, 0.0f, 5.0f});
    if (pos.size() == 3) position_ = glm::vec3(pos[0], pos[1], pos[2]);
    yaw_deg_    = j.value("yaw_deg",   yaw_deg_);
    pitch_deg_  = j.value("pitch_deg", pitch_deg_);
    auto tgt    = j.value("target", std::vector<float>{0.0f, 0.0f, 0.0f});
    if (tgt.size() == 3) target_ = glm::vec3(tgt[0], tgt[1], tgt[2]);
    if (j.contains("orbit")) {
        const auto& o = j["orbit"];
        orbit_distance_ = o.value("distance", orbit_distance_);
        orbit_yaw_      = o.value("yaw",      orbit_yaw_);
        orbit_pitch_    = o.value("pitch",    orbit_pitch_);
        orbit_speed_    = o.value("speed",    orbit_speed_);
    }
    if (j.contains("lens")) {
        const auto& l = j["lens"];
        fov_deg_ = l.value("fov_deg", fov_deg_);
        aspect_  = l.value("aspect",  aspect_);
        near_    = l.value("near",    near_);
        far_     = l.value("far",     far_);
    }
    if (j.contains("freefly")) {
        const auto& f = j["freefly"];
        move_speed_ = f.value("move_speed", move_speed_);
        look_speed_ = f.value("look_speed", look_speed_);
        boost_mul_  = f.value("boost_mul",  boost_mul_);
    }
}

}  // namespace gpusims

File: common/common-web/src/camera.ts
import { mat4, vec3 } from 'gl-matrix';

import type { JsonObject } from './types.js';
import type { InputSnapshot } from './input.js';

export type CameraMode = 'free-fly' | 'arcball' | 'orbit';

/** Re-exported for convenience; identical to InputSnapshot. */
export type CameraInputState = InputSnapshot;

const PITCH_LIMIT_DEG = 89.0;

export class Camera {
    mode: CameraMode = 'free-fly';

    // Free-fly state
    private _position: vec3 = vec3.fromValues(0, 0, 5);
    yawDeg = -90.0;       // looking down -Z
    pitchDeg = 0.0;

    /**
     * Camera position in world space. Read returns the underlying vec3
     * (in-place writes via gl-matrix functions are fine). Write accepts any
     * array-like of length ≥3 (vec3, number[], readonly tuple) and copies
     * into the underlying vec3 — preserves identity for code that holds
     * references to the live vec3.
     */
    get position(): vec3 { return this._position; }
    set position(v: ArrayLike<number>) {
        vec3.set(this._position, Number(v[0]), Number(v[1]), Number(v[2]));
    }

    // Arcball / Orbit state
    target: vec3 = vec3.fromValues(0, 0, 0);
    orbitDistance = 5.0;
    orbitYaw = 0.0;
    orbitPitch = 20.0;
    orbitSpeedDegPerSec = 30.0;

    // Lens
    fovDeg = 60.0;
    aspect = 16.0 / 9.0;
    near = 0.01;
    far = 1000.0;

    // Tuning
    moveSpeed = 5.0;     // units/s
    lookSpeed = 0.2;     // deg/pixel
    boostMultiplier = 5.0;

    update(dt: number, input: CameraInputState): void {
        switch (this.mode) {
            case 'free-fly': this.updateFreeFly(dt, input); break;
            case 'arcball':  this.updateArcball(input);     break;
            case 'orbit':    this.updateOrbit(dt);          break;
        }
    }

    private updateFreeFly(dt: number, input: CameraInputState): void {
        if (input.mouseRight) {
            this.yawDeg += input.mouseDx * this.lookSpeed;
            this.pitchDeg -= input.mouseDy * this.lookSpeed;
            this.pitchDeg = clamp(this.pitchDeg, -PITCH_LIMIT_DEG, PITCH_LIMIT_DEG);
        }
        const speed = this.moveSpeed * (input.shiftHeld ? this.boostMultiplier : 1.0);
        const fwd = this.forward();
        const rt  = this.right();

        const vel = vec3.fromValues(0, 0, 0);
        if (input.keyW) vec3.add(vel, vel, fwd);
        if (input.keyS) vec3.sub(vel, vel, fwd);
        if (input.keyD) vec3.add(vel, vel, rt);
        if (input.keyA) vec3.sub(vel, vel, rt);
        if (input.keyE) vec3.add(vel, vel, vec3.fromValues(0, 1, 0));
        if (input.keyQ) vec3.sub(vel, vel, vec3.fromValues(0, 1, 0));

        const len = vec3.length(vel);
        if (len > 0) {
            vec3.scale(vel, vel, (speed * dt) / len);
            vec3.add(this.position, this.position, vel);
        }
    }

    private updateArcball(input: CameraInputState): void {
        if (input.mouseLeft) {
            this.orbitYaw += input.mouseDx * this.lookSpeed;
            this.orbitPitch -= input.mouseDy * this.lookSpeed;
            this.orbitPitch = clamp(this.orbitPitch, -PITCH_LIMIT_DEG, PITCH_LIMIT_DEG);
        }
        if (input.scrollDy !== 0) {
            this.orbitDistance *= Math.pow(1.1, -input.scrollDy);
            this.orbitDistance = Math.max(this.orbitDistance, 0.001);
        }
        this.recomputeOrbitPosition();
    }

    private updateOrbit(dt: number): void {
        this.orbitYaw += this.orbitSpeedDegPerSec * dt;
        if (this.orbitYaw > 360.0) this.orbitYaw -= 360.0;
        this.recomputeOrbitPosition();
    }

    private recomputeOrbitPosition(): void {
        const yaw   = degToRad(this.orbitYaw);
        const pitch = degToRad(this.orbitPitch);
        const r = this.orbitDistance;
        const dx = r * Math.cos(pitch) * Math.cos(yaw);
        const dy = r * Math.sin(pitch);
        const dz = r * Math.cos(pitch) * Math.sin(yaw);
        vec3.set(this.position,
            this.target[0] + dx,
            this.target[1] + dy,
            this.target[2] + dz);
    }

    forward(): vec3 {
        const out = vec3.create();
        if (this.mode === 'free-fly') {
            const yaw = degToRad(this.yawDeg);
            const pitch = degToRad(this.pitchDeg);
            vec3.set(out,
                Math.cos(pitch) * Math.cos(yaw),
                Math.sin(pitch),
                Math.cos(pitch) * Math.sin(yaw));
            return vec3.normalize(out, out);
        }
        vec3.sub(out, this.target, this.position);
        return vec3.normalize(out, out);
    }

    right(): vec3 {
        const out = vec3.create();
        vec3.cross(out, this.forward(), vec3.fromValues(0, 1, 0));
        return vec3.normalize(out, out);
    }

    up(): vec3 {
        const out = vec3.create();
        vec3.cross(out, this.right(), this.forward());
        return vec3.normalize(out, out);
    }

    /** Column-major view matrix. Mirrors gpusims::Camera::view(). */
    view(): mat4 {
        const out = mat4.create();
        if (this.mode === 'free-fly') {
            const center = vec3.create();
            vec3.add(center, this.position, this.forward());
            return mat4.lookAt(out, this.position, center, vec3.fromValues(0, 1, 0));
        }
        return mat4.lookAt(out, this.position, this.target, vec3.fromValues(0, 1, 0));
    }

    /**
     * Column-major perspective matrix in WebGPU clip space ([0,1] depth, Y up
     * in NDC because of the y-flip). Identical to gpusims::Camera::projection
     * — Vulkan and WebGPU share clip-space convention.
     */
    projection(): mat4 {
        const out = mat4.create();
        mat4.perspectiveZO(out, degToRad(this.fovDeg), this.aspect, this.near, this.far);
        // Y-flip for Vulkan/WebGPU clip-space (origin top-left).
        out[5] *= -1;
        return out;
    }

    viewProjection(): mat4 {
        const out = mat4.create();
        mat4.multiply(out, this.projection(), this.view());
        return out;
    }

    setOrientation(yawDeg: number, pitchDeg: number): void {
        this.yawDeg = yawDeg;
        this.pitchDeg = clamp(pitchDeg, -PITCH_LIMIT_DEG, PITCH_LIMIT_DEG);
    }

    /**
     * Aim the camera at a world-space target. In free-fly mode this updates
     * yaw/pitch from the position→target direction so the camera looks at
     * (x,y,z); in arcball/orbit modes this also updates the orbit target.
     */
    lookAt(x: number, y: number, z: number): void {
        vec3.set(this.target, x, y, z);
        const dir = vec3.create();
        vec3.sub(dir, this.target, this._position);
        const len = vec3.length(dir);
        if (len < 1e-8) return;       // degenerate — leave orientation alone
        vec3.scale(dir, dir, 1 / len);
        // dir = (cos(p)*cos(y), sin(p), cos(p)*sin(y))
        const dy = clamp(dir[1], -1, 1);
        this.pitchDeg = (Math.asin(dy) * 180) / Math.PI;
        this.yawDeg = (Math.atan2(dir[2], dir[0]) * 180) / Math.PI;
        this.pitchDeg = clamp(this.pitchDeg, -PITCH_LIMIT_DEG, PITCH_LIMIT_DEG);
    }

    resetArcball(): void {
        this.orbitYaw = 0.0;
        this.orbitPitch = 20.0;
    }

    /** JSON serialization compatible with gpusims::Camera::toJson. */
    toJson(): JsonObject {
        const modeNum: Record<CameraMode, number> = { 'free-fly': 0, 'arcball': 1, 'orbit': 2 };
        return {
            mode: modeNum[this.mode],
            position: [this.position[0], this.position[1], this.position[2]],
            yaw_deg: this.yawDeg,
            pitch_deg: this.pitchDeg,
            target: [this.target[0], this.target[1], this.target[2]],
            orbit: {
                distance: this.orbitDistance,
                yaw: this.orbitYaw,
                pitch: this.orbitPitch,
                speed: this.orbitSpeedDegPerSec,
            },
            lens: {
                fov_deg: this.fovDeg,
                aspect: this.aspect,
                near: this.near,
                far: this.far,
            },
            freefly: {
                move_speed: this.moveSpeed,
                look_speed: this.lookSpeed,
                boost_mul: this.boostMultiplier,
            },
        };
    }

    fromJson(j: JsonObject): void {
        const modeNumToStr: Record<number, CameraMode> = {
            0: 'free-fly', 1: 'arcball', 2: 'orbit',
        };
        if (typeof j['mode'] === 'number') {
            this.mode = modeNumToStr[j['mode']] ?? 'free-fly';
        }
        const pos = j['position'];
        if (Array.isArray(pos) && pos.length === 3) {
            vec3.set(this.position, Number(pos[0]), Number(pos[1]), Number(pos[2]));
        }
        if (typeof j['yaw_deg']   === 'number') this.yawDeg   = j['yaw_deg'];
        if (typeof j['pitch_deg'] === 'number') this.pitchDeg = j['pitch_deg'];
        const tgt = j['target'];
        if (Array.isArray(tgt) && tgt.length === 3) {
            vec3.set(this.target, Number(tgt[0]), Number(tgt[1]), Number(tgt[2]));
        }
        const o = j['orbit'];
        if (o && typeof o === 'object' && !Array.isArray(o)) {
            if (typeof o['distance'] === 'number') this.orbitDistance       = o['distance'];
            if (typeof o['yaw']      === 'number') this.orbitYaw            = o['yaw'];
            if (typeof o['pitch']    === 'number') this.orbitPitch          = o['pitch'];
            if (typeof o['speed']    === 'number') this.orbitSpeedDegPerSec = o['speed'];
        }
        const l = j['lens'];
        if (l && typeof l === 'object' && !Array.isArray(l)) {
            if (typeof l['fov_deg'] === 'number') this.fovDeg = l['fov_deg'];
            if (typeof l['aspect']  === 'number') this.aspect = l['aspect'];
            if (typeof l['near']    === 'number') this.near   = l['near'];
            if (typeof l['far']     === 'number') this.far    = l['far'];
        }
        const f = j['freefly'];
        if (f && typeof f === 'object' && !Array.isArray(f)) {
            if (typeof f['move_speed'] === 'number') this.moveSpeed       = f['move_speed'];
            if (typeof f['look_speed'] === 'number') this.lookSpeed       = f['look_speed'];
            if (typeof f['boost_mul']  === 'number') this.boostMultiplier = f['boost_mul'];
        }
    }
}

function degToRad(deg: number): number { return (deg * Math.PI) / 180; }
function clamp(x: number, a: number, b: number): number { return x < a ? a : x > b ? b : x; }
